homeassistant/components/velux/cover.py

```diff
@@ -95,6 +97,33 @@ def is_closed(self) -> bool:
         """Return if the cover is closed."""
         return self.node.position.closed
 
+    @property
+    def is_opening(self) -> bool:
+        """Return if the cover is closing or not."""
+        return self.node.is_opening
+
+    @property
+    def is_closing(self) -> bool:
+        """Return if the cover is opening or not."""
+        return self.node.is_closing
+
+    @callback
+    def async_register_callbacks(self) -> None:
+        """Register callbacks to update hass after device was changed."""
+
+        async def after_update_callback(device) -> None:
+            """Call after device was updated."""
+            self.async_write_ha_state()
+            if self.node.is_moving():
+                if not self.is_looping_while_moving:
+                    self.is_looping_while_moving = True
+                    while self.node.is_moving():
+                        await asyncio.sleep(1)
+                        self.async_write_ha_state()
+                    self.is_looping_while_moving = False
+
+        self.node.register_device_updated_cb(after_update_callback)
```

Comment:
Why doesn't the node callback on every device update?

Reply:
The loop is used during device movement. From the API we get only sporadic position updates during the movement of covers, sometimes even none. However we receive a current position and target position information as well as an estimated completion datetime. Moreover once the cover has finished it's movement we receive a command finished notification. With that info we set the is_opening or is_closing and is_moving state. While is_moving we calculate the position to update the UI in a loop every second.

Reply:
When is a new position calculated?

https://github.com/Julius2342/pyvlx/blob/d1fee9e6041de6d73a8c90828cc33d6502af7541/pyvlx/opening_device.py#L170

Reply:
Ohh, thanks for that. I added a [new commit with that](https://github.com/home-assistant/core/pull/106684/commits/03a54791a6f86c3b837e2b2833b6c5f82670c680#diff-122894d3c6501f90d7be2e080a84f0823a54b52ae37908983e1087390a87791f).

Reply:
I'll check with the core team if we'll accept the loop as normally we don't allow calculated state in integrations that integrate devices or services from the Home Assistant side. If the library would keep the time itself, loop and callback etc, we wouldn't mind.

Reply:
I created a [PR ](https://github.com/Julius2342/pyvlx/pull/367/files) on pyvlx library, if this is the prefered way.

Reply:
@MartinHjelmare As far as I know HA Team is working on improvements of the cover platform. Finally I think it is not necessary during a movement to pull states periodically. More efficient would be an adjustment of the UI based on attributes of is_moving state to calculate the current position if someone is currently looking at the cover. This could be even more smoth than just update states every second and we wouldn't floot the states database. Considering this it is maybe better to have this loop now at HA component side than on library side. Let me know your opinion.

Reply:
The conclusion from the core team is that we don't want the loop in Home Assistant. The library can handle it if needed.